[FIX] web_editor: make o_not_editable contenteditable=false

Goaman opened the pull request:
The fix 1b80fe9 removed the behavior that made `.o_not_editable`
`contenteditable=false`. Without this behavior, other issues could
arises.





--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/87347).

Goaman commented:
@qsm-odoo 
> @Goaman This PR is not "urgent" but the fix (related to multiple PR and simply on its own) is. So if you can create a PR for your branch so I can have a look and merge, that would be nice +1 Also what about the other system merged by your team (I mentioned in [#82222 (comment)](https://github.com/odoo/odoo/pull/82222#discussion_r829173375))? Should it be kept alongside that o_not_editable behavior (I did not check in detail how it works)?
> 
> We can continue the discussion on the PR you will create, I guess it will be better +1 Thanks +1

The commit that you reference (https://github.com/odoo/odoo/pull/82222#discussion_r829173375) seems to keep contenteditable that are present before any modification of the editable (if any `contenteditable="..."` was in the source). The `contenteditable=false` that are added for `.o_not_editable` will not have the attribute `oe-keep-contenteditable` and will not be kept upon save.

rdeodoo commented:
FYI: https://github.com/odoo/odoo/pull/86825#pullrequestreview-926060662
Given the critical aspect of this `contenteditabe` property, we might as well ensure it won't break ever again :)
Probably something quite ugly as a tour that goes multiple time in edit mode to battle-test it :shrug:

Goaman commented:
> FYI: [#86825 (review)](https://github.com/odoo/odoo/pull/86825#pullrequestreview-926060662) Given the critical aspect of this `contenteditabe` property, we might as well ensure it won't break ever again :) Probably something quite ugly as a tour that goes multiple time in edit mode to battle-test it shrug

I've added a qunit test that should be enough. Tell me if you think I should add another kind of test.

rdeodoo commented:
I am not 100% sure what was decided here regarding the contenteditable attribute so I don't know if your test is enough or not.

I had a quick test and while it works fine for the contenteditable=false (indeed when entering edit mode, o_not_editable elements receives it), it doesn't work for contenteditable=true since we don't have a class for that?

I'd need a quick recap here about the scope we want to handle (after the fix ofc), tell me if something here is wrong:
1. manually setting contenteditable attribute on a template is forbidden / not supposed to be a thing / not supported
2. should be using only o_not_editable class
3. there is no class to add contenteditable=true (eg inside a contenteditable false), how are we supposed to do in such cases?

I just naively tried the following on your branch:
```xml
<div class="s_social_media text-left o_not_editable"> <!-- with or without contenteditable="false' it crashes -->
        <h4 class="s_social_media_title d-none" contenteditable="true">Social Media</h4>
        <a href="/website/social/facebook" class="s_social_media_facebook" target="_blank">
            <i class="fa fa-facebook rounded shadow-sm"/>
        </a>
        <a href="/website/social/twitter" class="s_social_media_twitter" target="_blank">
            <i class="fa fa-twitter rounded shadow-sm"/>
        </a>
        <a href="/website/social/linkedin" class="s_social_media_linkedin" target="_blank">
            <i class="fa fa-linkedin rounded shadow-sm"/>
        </a>
        <a href="/website/social/youtube" class="s_social_media_youtube" target="_blank">
            <i class="fa fa-youtube rounded shadow-sm"/>
        </a>
        <a href="/website/social/instagram" class="s_social_media_instagram" target="_blank">
            <i class="fa fa-instagram rounded shadow-sm"/>
        </a>
        <a href="/website/social/github" class="s_social_media_github" target="_blank">
            <i class="fa fa-github rounded shadow-sm"/>
        </a>
    </div>
    ```
 It crashed on `TypeError: element.querySelectorAll is not a function`
 
 Note that we have multiple case of contenteditable=true on element inside a contenteditable=false parent, to allow that element edition but no more than that. eg here, we want the user to be able to edit the title content, possibly adding stuff inside it (bold etc), but not mess with the DOM structure outside of that h4, so not possible to add a DOM element like an anchor next to that h4.
 
 That's why I am actually speaking a bout a ugly-naive tour in website that goes 2-3 times in edit mode / save to see if at the end on later edition, we have the DOM we expect to have.

rdeodoo commented:
By the way, even if you do something about contenteditable=true like "This is not allowed, we now remove those and use a new class for that (same way as for contenteditable=false)", keep in mind that some pages already have contenteditable=true in their page, so might as well ensure the code doesn't crash in that case (even if not officially supported and not in our codebase anymore if you remove those occurences in this PR (and its forward port!)

qsm-odoo commented:
> there is no class to add contenteditable=true (eg inside a contenteditable false), how are we supposed to do in such cases?

In fact, remembering how that worked before: we relied on `o_not_editable` puting the c`ontenteditable="false" `stuff automatically for the duration of edit mode, and for potential `contenteditable="true"` inside, it was the snippet option related to specific snippets which did it via their `start` method (and removing it in `destroy`). That's bad because you are not supposed to modify the DOM on options start... but kinda working in that case. Maybe we can introduce a class for that behavior or think about any other better way.

Goaman commented:
> I am not 100% sure what was decided here regarding the contenteditable attribute so I don't know if your test is enough or not.

There is no clear spec about it. When we replaced the editor, we inferred the spec.

> * manually setting contenteditable attribute on a template is forbidden / not supposed to be a thing / not supported

In which way it would be useful? Because if the `contenteditable="true"` is on the template, the user will be able to "edit" the content when he is not in edit mode. But if a good use case exists, we can support that behavior.

> * should be using only o_not_editable class

To have `contenteditable=false` when editing: yes.

> * there is no class to add contenteditable=true (eg inside a contenteditable false), how are we supposed to do in such cases?

To do it currently, add it manually when the editor does not observe the changes (between `observerActive` and `observerUnactive`).
If this is a frequent need, we can indeed introduce a new generic mechanism to do so (like a class (eg. `o_contenteditable`)).

> It crashed on `TypeError: element.querySelectorAll is not a function`

Indeed, I've fixed it.

> That's why I am actually speaking a bout a ugly-naive tour in website that goes 2-3 times in edit mode / save to see if at the end on later edition, we have the DOM we expect to have.

* Why would a tour provide a better test than a qunit? Do you have specific class of edge case in mind?
* Why edit/save more than 2 times would provide a better insurance? If after 2 times with the same instruction in between edit/save, the content is the same, how could the content be different a third time?

> By the way, even if you do something about contenteditable=true like "This is not allowed, we now remove those and use a new class for that (same way as for contenteditable=false)", keep in mind that some pages already have contenteditable=true in their page, so might as well ensure the code doesn't crash in that case (even if not officially supported and not in our codebase anymore if you remove those occurences in this PR (and its forward port!)

Removing `contenteditable=true` upon save already happen in `14.0` with the previous editor.

qsm-odoo commented on addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:
```diff
@@ -2130,7 +2144,36 @@ const Wysiwyg = Widget.extend({
+        const nodes = (element && element.querySelectorAll && element.querySelectorAll('.o_not_editable:not([contenteditable=false])')) || [];
+        for (const node of nodes) {
+            node.setAttribute('contenteditable', false);
+            let observer = this._oNotEditableObservers.get(node);
```
Here are the occurences of your new map:
![image](https://user-images.githubusercontent.com/10338094/170280519-ca8fa1a2-1585-4861-8b07-314bdb20d6b7.png)

Creation, loop though, get and delete... where do you actually put something in that map? I must be dumb and missing something obvious :grimacing:

Goaman commented on addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:
```diff
@@ -695,7 +702,7 @@ export class OdooEditor extends EventTarget {
             return;
         }
-        this.observerFlush();
+        this.sanitize();
```
> I was questioning the fact we now sanitize during `historyStep`, that seems strange to me considering not other sanitization call seems to be removed in this PR and especially without explanations for the reader to understand :shrug: 

The main problem was that there was not sanitization when an automatic step was made (see `this.observerTimeout`).
Which made me think: when do we want to make an historyStep without sanitizing? Maybe when we know that nothing is to be sanitized like in undo/redo. But as the sanitization is not performance heavy, we can simply sanitize each time in order to avoid forgetting to sanitize before an historyStep.

> So I went and have a look where sanitize is currently called, I see this:
> ![image](https://user-images.githubusercontent.com/10338094/170277416-27aadcb5-f94f-473b-9b9b-d1fe156373cf.png)
> 
> So apparently when reaching that code, we sanitize twice? shrug

The double sanitization was a mistake, I've removed the `this.sanitize();this.historystep();`.

> Clearly there is more to think about here and comments to explain what's going on, that's not supposed to reach the review step. 

What is it supposed to reach? Isn't it the role of the reviewer to spot problem that the developer didn't foresee?

> Even your commit message (mentioning the wrong apps) is clearly too small considering all the discussions that went around this problem we are trying to solve.

Indeed, there is one module mistake in the commit, thank you for spotting it.
For me, the discussion is simple: there was a behavior before the new editor that we broke and that need a fix otherwise the part of the code that require that behavior does not work. If we had fixed it in the editor merge, no discussions would have ever been referenced.

> Probably also more things to consider as explained here [#85227 (comment)](https://github.com/odoo/odoo/pull/85227#discussion_r880401952) (at least as comment addition but probably also unifying the behavior of `o_not_editable` and `getReadonlyAreas`?)
I'm suggesting to revert the commit of #85227 in master as you do not want to do it in 15.0. (I've tested and the fix #85227 became useless as some point as I can't reproduce it when reverting it in 15.0 (without this PR)).

qsm-odoo replied:
> The double sanitization was a mistake, I've removed the this.sanitize();this.historystep();.

 >>   Clearly there is more to think about here and comments to explain what's going on, that's not supposed to reach the review step.

> What is it supposed to reach? Isn't it the role of the reviewer to spot problem that the developer didn't foresee?

Exactly: spot problems that the developer did not foresee. I don't want to be mean but there was nothing to "foresee" here :grimacing:  You add a `sanitize` in `historyStep`:
- You thus grep `sanitize` to see if any useless call now remains -> yes the one next to a call to `historyStep`
- You thus grep `historyStep` to see if you actually wanted to sanitize each time it is called -> you also find the `sanitize` calll next to `historyStep`

-> You forgot to do this during development? Not a problem, you then self-review your PR, see a `sanitize` call added without comment and not any other call removed, you go back to your code to check.

-> You missed that during your self-review? Not a problem either, it's a mistake, everyone make some but, as I said, that's not my role as a reviewer to spot **that kind** of mistake (more often that not, **I will miss those and they will get merged**).

Again, I really don't want to be mean. You apparently did not like me saying this should not reach the review step, but that was not about the `sanitize` call only. It was paired with the unwritten map I mentioned below https://user-images.githubusercontent.com/10338094/170280519-ca8fa1a2-1585-4861-8b07-314bdb20d6b7.png, the huge lack of comments in both the code and the commit message, ... I really got a real vibe of "Ok, it should work, let's send to QSM, maybe he will r+"... :weary: 

Reviewing this kind of PR (spotting that kind of issues) takes more time it would take me to do the actual development:
- "Ok so a call to `sanitize` there... why...? Don't know... let's grep, mmh there are useless calls here and there"
- "Ok so a map... why and how is it used... let's add a debugger to try and understand"
- etc

> If we had fixed it in the editor merge, no discussions would have ever been referenced.

Are you actually saying that if we developed without bugs, we would not need bug fixes and thus we don't need to explain bug fixes? :s 

> I'm suggesting to revert the commit of https://github.com/odoo/odoo/pull/85227 in master [...]

From my understanding, @dmo-odoo wanted to keep the functionality and as I explained I have nothing against it but it would require unifications in behaviors IMO. I let you decide, I am fine closing my eyes until a bug is reported if you prefer (as a bug is actually unlikely I guess, it's more about having robust behaviors).

> [...] as you do not want to do it in 15.0.

More the stable policy than me :wink: If you judge the code is recent enough to break stable policy, revert in stable and r+, I am fine closing my eyes too, that's just not something I would do myself.

Goaman commented on addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:
```diff
@@ -2130,7 +2144,36 @@ const Wysiwyg = Widget.extend({
         this.ptp.notifyAllClients('ptp_join');
+    },
+    _setONotEditable: function (element) {
+        const nodes = (element && element.querySelectorAll && element.querySelectorAll('.o_not_editable:not([contenteditable=false])')) || [];
```
I've added a jsdoc.

dmo-odoo commented:
@robodoo r+
